Bioinformatics QC page & simplfy analysis finalise

erikwaskiewicz opened the pull request:
# Minor Software Update 

### Description 
- Change the drop down where sample analyses were closed to have a series of drop downs so that less mistakes are made
  - Added send back button with breakglass
  - Close sample button now pops up a modal
  - Modal will do checks that used to happen on submit as a ajax calls, will only carry on to next step if the checks pass/ will limit options if they fail
- Added a bioinformatics QC step before results are released
  - idea is that they can be auto uploaded here but only visible to scientists once QC is done
  - added permission so only bioinf team can do the QC
  - Added form to duplicate info from AutoQC, in future would like to automate this
  - Added bioinf QC form to simplify the bioinf QC step
  - This hasnt been fully tested so might be a bit clunky, might try to drop the patient name at the same time as deploying this to simplify further
- Simplified worksheet status so it'll hopefully be quicker to display

Outstanding:
- tests have been broken - fixtures changes and more model functions mean logic has moved, will need overhaul
- Hasnt had really detailed testing, particularly bioinf QC bits


### Justification 

Not changing the logic, just how the database functions
  

_______________________________________________________________________________________________________

## Testing

### Description :

Ongoing, see above, just an interim code review please

### Data Location:

N/A


_______________________________________________________________________________________________________

## Checklist
< Place an X in the appropriate boxes, if not applicable state N/A >

- [ ] Is the test data easily accesible?  
- [ ] Has the pipeline/software been run end to end by the tester?
- [ ] Has the pipeline/software been code reviewed on github?  
- [ ] Has the pipeline/software version number been incremented?
- [ ] Are all relevant doccuments updated e.g. SOP, ReadME, Pull request 

Additional checklist for pipeline updates (if applicable)
- [ ] Is the sensitivity or other key performance metrics unaffected by the changes? 
- [ ] If the environment changes has this been deployed and tested in place?
- [ ] Has the pipeline version in the variables file been updated e.g. development > master

Istaisa commented:
Doing this review as a comment per chunk of testing.
**User testing - bioinformatics QC checks** (note have not checked underlying code yet)

**User Groups / Display**
- Users not in the QC group can see what needs bioinformatics QC but cannot click on to the runs
- Users in the QC group can click on the runs and perform QC
COMMENTS/ISSUES
- Add instructions to the docs on how to set up the QC user group locally

**Performing QC Checks**
- I really like this section I think it's going to make failing more straightforward and querying better in the future
- Have tried combinations of run pass all samples pass, run pass some samples fail, whole run fail
COMMENTS/ISSUES
- Can't edit the patient name within the bioinformatics QC page (but losing it altogether might be the easiest way to solve this)
- You can use the same AutoQC primary key for different runs  - do we want this or should we force it to be unique?
- AutoQC URL needs changing: should be 10.59.210.245:5000/**run_analysis**/<pk>
- Samples aren't finalising properly from the bioinformatics QC checks. The status in the table changes from bioinformatics check 2 to bioinformatics QC fail, but when you go in to the sample you've still got the check interface open and can't download the report/XML

Istaisa commented:
**User testing - updated pass/fail interface** (note have not checked any of the underlying code yet)

**General Interface**
- Can't finalise at all without checking patient demographics and updating name
- Can't pass RNA without checking all the fusions 
- Can't pass DNA without doing all the DNA checks
- Send back check button works as expected 

**PASSING SAMPLES**
- If everything agrees after 2 checks it closes nicely
- If last 2 checks agree after 3 checks it closes nicely
COMMENTS/ISSUES
- Will finalise if user 1 and user 3 are the same person, or if user 2 and user 3 are the same person which means it can finalise with some variants only having been checked by 1 person

**FAILING**
- The status of a sample awaiting a fail is now 'IGV check 2' and the button's less obvious - this might be the intended behaviour though?
- 2 fails works correctly
- pass then fail then fail works correctly (yay!)
- fail then pass then pass works correctly
- fail then pass then fail then fail works BUT checker 3 and 4 can be the same person (see above - same thing with 2 checks rather than 2 checks by independent users)
- scientists can't reopen a completed fail sample - i can't remember if we only added that functionality in on complete samples for a reason

COMMENTS/ISSUES
- Overall very clean but would be good if it was 2 independent matching checks not just 2 matching checks
- This is more of a GUI one - but would it be possible to make some of the tickboxes a little bit bigger? e.g. on the demographics page tickbox. Lots of dexterity required on a big screen!

Istaisa commented on analysis/forms.py:
```diff
@@ -53,6 +53,62 @@ def __init__(self, *args, **kwargs):
+
+
+class SampleQCForm(forms.Form):
+    """
```
is this just binfx or scientist or both? add to docstring

erikwaskiewicz replied:
Just bioinformatics, updated docstring

Istaisa commented on analysis/forms.py:
```diff
@@ -118,23 +174,52 @@ class SubmitForm(forms.Form):
     Finalise a sample analysis
```
edit docstring

erikwaskiewicz replied:
Done

Istaisa commented on analysis/forms.py:
```diff
@@ -118,23 +174,52 @@ class SubmitForm(forms.Form):
+
+class SendCheckBackForm(forms.Form):
+    """
+    Finalise a sample analysis
```
edit docstring

Istaisa commented on analysis/forms.py:
```diff
@@ -118,23 +174,52 @@ class SubmitForm(forms.Form):
 
+class DetailsCheckForm(forms.Form):
+    """
+    Finalise a sample analysis
```
edit docstring

Istaisa commented on analysis/models.py:
```diff
@@ -161,64 +199,287 @@ def percent_reads_ntc(self):
+        if self.worksheet.auto_qc_pass_fail == 'F':
+            return 'Bioinformatics QC fail'
+        elif self.sample_pass_fail != '-':
+            return self.get_sample_pass_fail_display()
```
does this function exist?

erikwaskiewicz replied:
It's some Django magic, any model field with choices will display the pretty printed version with the built in Django get_*_display() function. Will add a comment to code

Istaisa commented on analysis/models.py:
```diff
@@ -161,64 +199,287 @@ def percent_reads_ntc(self):
+    def get_checks(self):
+        """
+        Get all associated checks and work out the status
+        TODO - this is getting called twice when samples page is loaded - think its an AJAX issue
```
looks like this is only being called once now based on adding a print statement

erikwaskiewicz replied:
Agree - same on mine too, will remove comment

Istaisa commented on analysis/models.py:
```diff
@@ -161,64 +199,287 @@ def percent_reads_ntc(self):
+    def all_panel_fusions(self):
+        return FusionPanelAnalysis.objects.filter(sample_analysis=self)
+
+    def snvs_have_2_checks(self):
```
As this and the fusion functions have the same logic it might be better as one function with a variant/fusion option? or rename the variables in the fusion function to be all_fusion_checks / for f in self.all_panel_fusions()?

erikwaskiewicz replied:
Made into one function

Istaisa commented on analysis/models.py:
```diff
@@ -161,64 +199,287 @@ def percent_reads_ntc(self):
+
+        # make check objects for all fusions
+        if self.panel.show_fusions:
+            for v in self.all_panel_fusions():
```
can we change v to f here

erikwaskiewicz replied:
Done

Istaisa commented:
My unit tests do work - don't know if that's concerning if they're meant to have broken

Istaisa reviewed (commented):
Not done proper review of the actual code yet but done user testing for QC and checks and identified the bits that are(n't) working

erikwaskiewicz commented:
All comments in code have been addressed

Comments from testing:
- *Add instructions to the docs on how to set up the QC user group locally* - Done
- *Can't edit the patient name within the bioinformatics QC page (but losing it altogether might be the easiest way to solve this)* - Done, removed patient name altogether
- *You can use the same AutoQC primary key for different runs - do we want this or should we force it to be unique?* - Done, just added unique=True. In future would look at linking better with AutoQC
- *AutoQC URL needs changing: should be 10.59.210.245:5000/run_analysis/* - Done, changed to webserver URLs
- *Samples aren't finalising properly from the bioinformatics QC checks. The status in the table changes from bioinformatics check 2 to bioinformatics QC fail, but when you go in to the sample you've still got the check interface open and can't download the report/XML* - Done, needed to tweak loops in analysis page and formatting of PDF download
- *The status of a sample awaiting a fail is now 'IGV check 2' and the button's less obvious - this might be the intended behaviour though?* - this was kind intended, going to go with it and see if they complain (in theory should be a fresh check each time)
- *scientists can't reopen a completed fail sample - i can't remember if we only added that functionality in on complete samples for a reason* - I think they can, as long as they're logged in?
- *would be good if it was 2 independent matching checks not just 2 matching checks* - agree but not got to it, will add as issue (it does make testing a little more faffy because you have to log in and out lots, so will leave til last)
- *This is more of a GUI one - but would it be possible to make some of the tickboxes a little bit bigger?* -  had a quick go but didnt work, not sure how as crispy forms handling it, will add as issue

Other issues:
- Fixed issue where after reopening sample pass/fail is "P" instead of "-" (will also need a plan to handle this when deploying as open Checks will have status of "P", will have to do a query to replace with "-", auto_qc_pass_fail on Worksheet model also like this, comment in code re: migrations on deployment, will need to do a test run)
- Added option for bioinformatics to send a check back
- removed this TODO from the code, will add as issue - "'not analysed' option for sample/run, would need a generic 'not analysed' panel probably"

Outstanding:
- unit tests still need work, quite a big overhaul with functions split into smaller chunks

Istaisa reviewed (changes requested):
Checked that the requested changes in the comment above make sense and work correctly - all good. Will do final review when unit tests updated of models function logic etc.
